nx/lib/nx/lin_alg.ex

```diff
@@ -1117,38 +1117,59 @@ defmodule Nx.LinAlg do
         48.0
       >
   """
-  def determinant(tensor)
+  defn determinant(tensor) do
+    Nx.Defn.Kernel.assert_shape_pattern(tensor, {n, n})
+
+    {n, _} = Nx.shape(tensor)
+
+    cond do
+      Nx.equal(n, 2) ->
+        determinant_2by2(tensor)
+
+      Nx.equal(n, 3) ->
+        determinant_3by3(tensor)
+
+      :otherwise ->
+        determinant_NbyN(tensor)
+    end
+  end
 
   # for 2x2 and 3x3, use the algebraic closed formula
-  def determinant(%{shape: {2, 2}} = t) do
+  defnp determinant_2by2(t) do
     t = Nx.tile(t, [1, 2])
 
     t
     |> diagonal_product(0)
     |> Nx.subtract(diagonal_product(t, 1))
   end
 
-  def determinant(%{shape: {3, 3}} = t) do
+  defnp determinant_3by3(t) do
     pos_t = Nx.tile(t, [1, 2])
 
     neg_t = Nx.reverse(pos_t, axes: [1])
 
-    Enum.reduce(0..2, Nx.tensor(0), fn offset, acc ->
-      acc
-      |> Nx.add(diagonal_product(pos_t, offset))
-      |> Nx.subtract(diagonal_product(neg_t, offset))
-    end)
+    # This is an unrwapped Enum.reduce(0..2, Nx.tensor(0), fn offset, acc -> .. end)
+
+    Nx.tensor(0)
+    |> Nx.add(diagonal_product(pos_t, 0))
+    |> Nx.subtract(diagonal_product(neg_t, 0))
+    |> Nx.add(diagonal_product(pos_t, 1))
+    |> Nx.subtract(diagonal_product(neg_t, 1))
+    |> Nx.add(diagonal_product(pos_t, 2))
+    |> Nx.subtract(diagonal_product(neg_t, 2))
```

Comment:
The compiler complained about Enum.reduce not being defined as defn, so I ended up unwrapping it since there were few iterations. Should I use while instead?

Reply:
Nah, this is fine as is. You could also use a transform.